Parallel Paths

misterhat opened the pull request:
add `.setParallel(limit)`

when many varying path lengths are used, distributes the path finding per iterationsPerCalculation on up to limit paths. when calculate() is called, switch to a solving a different path in the queue instead of solving them sequentially.

if the parallel limit is 1, easystar acts as it currently does by resolving each path in the order they were submitted. if it's set to 2, it will switch between two path finding instances every `.calculate()` call. if -1, switch the path finding instance every time `.calculate()` is called.

it still results in the same total amount of time to calculate the paths, but the load distribution could be useful.

e.g.
```javascript

const EasyStar = require('easystarjs');               
                                                      
const easystar = new EasyStar.js();                   
const grid = require('./obstacles.json') ;            
grid[1162][938] = 0;                                  
easystar.setGrid(grid);                               
easystar.setAcceptableTiles([0]);
easystar.setIterationsPerCalculation(500);            
easystar.setParallelLimit(-1); // 1, 2, etc.

for (let i = 0; i < 5; i += 1) {                      
    console.time(`long-path-${i}`);                   
    easystar.findPath(1380, 1318, 368, 982, () => {   
        console.timeEnd(`long-path-${i}`);            
    });                                               
}                                                     

for (let i = 0; i < 50; i += 1) {                     
    console.time(`short-path-${i}`);
    easystar.findPath(1380, 1318, 1360, 1324, () => {
        console.timeEnd(`short-path-${i}`);
    });
}

const calculate = () => {                             
    easystar.calculate();                             
    setTimeout(calculate, 5);                         
};                                                    
calculate();
```

with setParallelLimit unset or set to 1, it resolves the long-paths first (at ~400ms each), then the short-paths after ~2 seconds. with setParallelLimit set to -1, it resolves the short-paths from 60-120ms and the long paths after ~2 seconds.

this is useful if you have small paths mixed with large paths, as the small paths can return much quicker even if they were added during the solving of a much larger and slower path.

prettymuchbryce commented on bin/easystar-0.4.3.js:
```diff
@@ -78,9 +78,11 @@ var EasyStar =
 	    var iterationsSoFar;
 	    var instances = {};
 	    var instanceQueue = [];
+	    var instanceQueueIndex = 0;
```
Don't change the files in `bin/*`. They will be compiled when the version is bumped.

misterhat replied:
will do. i'll checkout the earlier revision.

prettymuchbryce commented on src/easystar.js:
```diff
@@ -178,6 +180,19 @@ EasyStar.js = function() {
+    * @param {Number} limit The maximum number of iterations before resetting
+    * on calculate() call. Use -1 to run all the paths at once.
+    **/
+    this.setParallelLimit = function(limit) {
```
I'm trying to think of a use case where it makes sense to use anything other than `-1` or `1` as an argument to `setParallelLimit`. What do you think about making this something like `enableParallelCompute` and `disableParallelCompute`? This follows in line more with the `enableSync` option.

In v2 I'm thinking we should start accepting options like this in the constructor of `EasyStar` rather than these ad-hoc functions.

misterhat replied:
if I submit 10 1 second paths and set the limit to -1, it will take ~10 seconds before it returns all the paths at once. if i set the limit to 2, it will return one path every 2 seconds instead. i also didn't need this feature, but I thought it could be useful to get paths returned quicker and still solve more than one, but maybe it's not worth being exposed here? I can replace it with the binary option for now.

prettymuchbryce commented on src/easystar.js:
```diff
@@ -544,6 +566,19 @@ EasyStar.js = function() {
+
+        instanceQueueIndex--;
+
+        if (instanceQueueIndex < 0 || instanceQueueIndex > instanceQueue.length) {
```
Can you help me understand why this condition would ever be true here?

`instanceQueueIndex > instanceQueue.length`

It seems like we're always decrementing the instanceQueueIndex?

misterhat replied:
i believe you're right. i originally had it increment instead of decrement, so I'll remove that check.

misterhat commented:
Ok I think this is what you asked for? I restored the original bin/* files from the master branch, changed the API + typescript bindings and added unit tests for serial and parallel pathfinding.
